iOS 11/Xcode 9 compatibility

# Expected behavior
To be able to deploy to the iOS simulator when Xcode 9 is installed with the iOS 11 SDK, with the ability to select the additional simulators from the list.

# Actual behavior.
When I try to run `cordova emulate ios` without specifying a default device, I receive the following error:

    No target specified for emulator. Deploying to undefined simulator

    Device type "com.apple.CoreSimulator.SimDeviceType.undefined" could not be found.
    [ERROR] An error occurred while running cordova emulate ios (exit code 1).

Despite there being default simulators installed, it appears like it is not selecting the first relevant simulator, instead, just whatever is at the top of the list.

If I run `ios-sim showdevicetypes`, I am returned the following list, which is omitting the new simulators (iPhone 8, iPhone 8 Plus and the iPhone X), and has the Apple-Watch listed at the start (which, as explained below, is now the default):

    Apple-Watch-38mm, watchOS 4.0
    Apple-Watch-42mm, watchOS 4.0
    Apple-Watch-Series-2-38mm, watchOS 4.0
    Apple-Watch-Series-2-42mm, watchOS 4.0
    iPhone-5s, 11.0
    iPhone-6, 11.0
    iPhone-6-Plus, 11.0
    iPhone-6s, 11.0
    iPhone-6s-Plus, 11.0
    iPhone-7, 11.0
    iPhone-7-Plus, 11.0
    iPhone-SE, 11.0
    iPad-Air, 11.0
    iPad-Air-2, 11.0
    iPad--5th-generation-, 11.0
    iPad-Pro--12-9-inch---2nd-generation-, 11.0
    iPad-Pro--10-5-inch-, 11.0
    Apple-TV-1080p, tvOS 11.0
    Apple-TV-4K-4K, tvOS 11.0
    Apple-TV-4K-1080p, tvOS 11.0
    iPad-Pro--9-7-inch-, 11.0
    iPad-Pro, 11.0

If I then run `ios-sim start`, it states:

    --devicetypeid was not specified, using first available device: Apple Watch - 38mm.

I expect, because of this, it's trying to deploy the Cordova app to the Apple Watch, hazard a guess?

If I _do_ specify an iOS emulator to deploy to, I receive the following:

    $ ionic cordova emulate ios --target="iPhone-7"
    ...
    No available runtimes could be found for "iPhone 7".
    [ERROR] An error occurred while running cordova emulate ios --target iPhone-7 (exit code 1).

# Steps to reproduce the problem
As above (assuming Cordova and ios-sim are installed).

1. Start a blank Cordova project: `cordova start SimApp && cd SimApp`
2. Run the app: `cordova emulate ios --target="iPhone-7"`

# System Specs

Please run the commands below in your Terminal.app and include it in the issue. Check when done and include results below.

- [x] 1. system_profiler SPSoftwareDataType
- [x] 2. ios-sim --version
- [x] 3. xcodebuild -version
- [x] 4. xcode-select --print-path
- [x] 5. node --version
- [x] 6. npm --version

**1. system_profiler SPSoftwareDataType:**

    System Software Overview:

      System Version: macOS 10.12.6 (16G29)
      Kernel Version: Darwin 16.7.0
      Boot Volume: Macintosh HD
      Boot Mode: Normal
      Computer Name: Kelvin’s MacBook Pro
      User Name: Kelvin Dart (kelvindart)
      Secure Virtual Memory: Enabled
      System Integrity Protection: Enabled
      Time since boot: 1 day 6:13

**2. ios-sim --version:**

    6.0.0

**3. xcodebuild -version:**

    Xcode 9.0
    Build version 9A235

**4. xcode-select --print-path:**

    /Applications/Xcode.app/Contents/Developer

**5. node --version:**

    v8.2.1

**6. npm --version:**

    5.4.1

--- a/src/lib.js
+++ b/src/lib.js
@@ -265,6 +265,23 @@ function filterDeviceName(deviceName) {
     return deviceName;
 }
 
+// Xcode 9 `xcrun simctl list devicetypes` have obfuscated names for 2017 iPhones and Apple Watches.
+// `iPad Pro` in iOS 9.3 has mapped to `iPad Pro (9.7 inch)`
+// `Apple TV 1080p` has mapped to `Apple TV`
+function deviceNameMapFix(deviceName) {
+
+    var m = {
+        "iPhone 8": "iPhone2017-A",
+        "iPhone 8 Plus": "iPhone2017-B",
+        "iPhone X": "iPhone2017-C",
+        "Apple Watch Series 3 - 38mm": "Watch2017 - 38mm",
+        "Apple Watch Series 3 - 42mm": "Watch2017 - 42mm",
+        "Apple TV 1080p" : "Apple TV",
+        "iPad Pro" : "iPad Pro (9.7-inch)"
+    };
+    return m[deviceName];
+}
+
 var lib = {
 
     init: function() {
@@ -327,7 +344,12 @@ var lib = {
             var dname = filterDeviceName(deviceName);
 
             if (!(dname in name_id_map)) {
-                continue;
+                var fixedname = deviceNameMapFix(dname);
+                if (fixedname) {
+                    dname = filterDeviceName(fixedname);
+                } else {
+                    continue;
+                }
             }
 
             runtimes.forEach(cur(dname));

--- a/src/lib.js
+++ b/src/lib.js
@@ -141,6 +141,7 @@ function getDeviceFromDeviceTypeId(devicetypeid) {
 
     var options = { 'silent': true };
     var list = simctl.list(options).json;
+    list = fixSimCtlList(list);
 
     var arr = [];
     if (devicetypeid) {
@@ -279,6 +280,30 @@ function fixNameKey(array, mapping) {
     });
 }
 
+function fixSimCtlList(list) {
+    // Xcode 9 `xcrun simctl list devicetypes` have obfuscated names for 2017 iPhones and Apple Watches.
+    var deviceTypeNameMap = {
+        'iPhone2017-A': 'iPhone 8',
+        'iPhone2017-B': 'iPhone 8 Plus',
+        'iPhone2017-C': 'iPhone X',
+        'Watch2017 - 38mm': 'Apple Watch Series 3 - 38mm',
+        'Watch2017 - 42mm': 'Apple Watch Series 3 - 42mm'
+    };
+    list.devicetypes = fixNameKey(list.devicetypes, deviceTypeNameMap);
+
+    // `iPad Pro` in iOS 9.3 has mapped to `iPad Pro (9.7 inch)`
+    // `Apple TV 1080p` has mapped to `Apple TV`
+    var deviceNameMap = {
+        'Apple TV 1080p': 'Apple TV',
+        'iPad Pro': 'iPad Pro (9.7-inch)'
+    };
+    Object.keys(list.devices).forEach(function(key) {
+        list.devices[key] = fixNameKey(list.devices[key], deviceNameMap);
+    });
+
+    return list;
+}
+
 var lib = {
 
     init: function() {
@@ -301,6 +326,7 @@ var lib = {
     showsdks: function(args) {
         var options = { silent: true, runtimes: true };
         var list = simctl.list(options).json;
+        list = fixSimCtlList(list);
 
         console.log('Simulator SDK Roots:');
         list.runtimes.forEach(function(runtime) {
@@ -316,28 +342,7 @@ var lib = {
     getdevicetypes: function(args) {
         var options = { silent: true };
         var list = simctl.list(options).json;
-
-        // Xcode 9 `xcrun simctl list devicetypes` have obfuscated names for 2017 iPhones and Apple Watches.
-        var deviceTypeNameMap = {
-            'iPhone2017-A': 'iPhone 8',
-            'iPhone2017-B': 'iPhone 8 Plus',
-            'iPhone2017-C': 'iPhone X',
-            'Watch2017 - 38mm': 'Apple Watch Series 3 - 38mm',
-            'Watch2017 - 42mm': 'Apple Watch Series 3 - 42mm'
-        };
-        list.devicetypes = fixNameKey(list.devicetypes, deviceTypeNameMap);
-
-        // `iPad Pro` in iOS 9.3 has mapped to `iPad Pro (9.7 inch)`
-        // `Apple TV 1080p` has mapped to `Apple TV`
-        var deviceNameMap = {
-            'Apple TV 1080p': 'Apple TV',
-            'iPad Pro': 'iPad Pro (9.7-inch)'
-        };
-        Object.keys(list.devices).forEach(function(key) {
-            list.devices[key] = fixNameKey(list.devices[key], deviceNameMap);
-        });
-
-        //console.log(list.devices['tvOS 11.0']);
+        list = fixSimCtlList(list);
 
         var druntimes = findRuntimesGroupByDeviceProperty(list, 'name', true);
         var name_id_map = {};

--- a/src/lib.js
+++ b/src/lib.js
@@ -265,21 +265,18 @@ function filterDeviceName(deviceName) {
     return deviceName;
 }
 
-// Xcode 9 `xcrun simctl list devicetypes` have obfuscated names for 2017 iPhones and Apple Watches.
-// `iPad Pro` in iOS 9.3 has mapped to `iPad Pro (9.7 inch)`
-// `Apple TV 1080p` has mapped to `Apple TV`
-function deviceNameMapFix(deviceName) {
+function fixNameKey(array, mapping) {
+    if (!array || !mapping) {
+        return array;
+    }
 
-    var m = {
-        'iPhone 8': 'iPhone2017-A',
-        'iPhone 8 Plus': 'iPhone2017-B',
-        'iPhone X': 'iPhone2017-C',
-        'Apple Watch Series 3 - 38mm': 'Watch2017 - 38mm',
-        'Apple Watch Series 3 - 42mm': 'Watch2017 - 42mm',
-        'Apple TV 1080p': 'Apple TV',
-        'iPad Pro': 'iPad Pro (9.7-inch)'
-    };
-    return m[deviceName];
+    return array.map(function(elem){
+        var name = mapping[elem.name];
+        if (name) {
+            elem.name = name;
+        }
+        return elem;
+    });
 }
 
 var lib = {
@@ -320,6 +317,28 @@ var lib = {
         var options = { silent: true };
         var list = simctl.list(options).json;
 
+        // Xcode 9 `xcrun simctl list devicetypes` have obfuscated names for 2017 iPhones and Apple Watches.
+        var deviceTypeNameMap = {
+            'iPhone2017-A': 'iPhone 8',
+            'iPhone2017-B': 'iPhone 8 Plus',
+            'iPhone2017-C': 'iPhone X',
+            'Watch2017 - 38mm': 'Apple Watch Series 3 - 38mm',
+            'Watch2017 - 42mm': 'Apple Watch Series 3 - 42mm'
+        };
+        list.devicetypes = fixNameKey(list.devicetypes, deviceTypeNameMap);
+
+        // `iPad Pro` in iOS 9.3 has mapped to `iPad Pro (9.7 inch)`
+        // `Apple TV 1080p` has mapped to `Apple TV`
+        var deviceNameMap = {
+            'Apple TV 1080p': 'Apple TV',
+            'iPad Pro': 'iPad Pro (9.7-inch)'
+        };
+        Object.keys(list.devices).forEach(function(key) {
+            list.devices[key] = fixNameKey(list.devices[key], deviceNameMap);
+        });
+
+        //console.log(list.devices['tvOS 11.0']);
+
         var druntimes = findRuntimesGroupByDeviceProperty(list, 'name', true);
         var name_id_map = {};
 
@@ -344,12 +363,10 @@ var lib = {
             var dname = filterDeviceName(deviceName);
 
             if (!(dname in name_id_map)) {
-                var fixedname = deviceNameMapFix(dname);
-                if (fixedname) {
-                    dname = filterDeviceName(fixedname);
-                } else {
-                    continue;
-                }
+
+                console.log('----> DNAME', dname);
+                console.log('----> RUNTIMES', runtimes);
+                continue;
             }
 
             runtimes.forEach(cur(dname));

--- a/src/lib.js
+++ b/src/lib.js
@@ -271,13 +271,13 @@ function filterDeviceName(deviceName) {
 function deviceNameMapFix(deviceName) {
 
     var m = {
-        "iPhone 8": "iPhone2017-A",
-        "iPhone 8 Plus": "iPhone2017-B",
-        "iPhone X": "iPhone2017-C",
-        "Apple Watch Series 3 - 38mm": "Watch2017 - 38mm",
-        "Apple Watch Series 3 - 42mm": "Watch2017 - 42mm",
-        "Apple TV 1080p" : "Apple TV",
-        "iPad Pro" : "iPad Pro (9.7-inch)"
+        'iPhone 8': 'iPhone2017-A',
+        'iPhone 8 Plus': 'iPhone2017-B',
+        'iPhone X': 'iPhone2017-C',
+        'Apple Watch Series 3 - 38mm': 'Watch2017 - 38mm',
+        'Apple Watch Series 3 - 42mm': 'Watch2017 - 42mm',
+        'Apple TV 1080p': 'Apple TV',
+        'iPad Pro': 'iPad Pro (9.7-inch)'
     };
     return m[deviceName];
 }

--- a/src/lib.js
+++ b/src/lib.js
@@ -270,7 +270,7 @@ function fixNameKey(array, mapping) {
         return array;
     }
 
-    return array.map(function(elem){
+    return array.map(function(elem) {
         var name = mapping[elem.name];
         if (name) {
             elem.name = name;

